src/ResourceGroupParser.h

```diff
@@ -183,22 +237,63 @@ namespace snowcrash {
             return SectionProcessorBase<ResourceGroup>::isUnexpectedNode(node, sectionType);
         }
 
-        /** Finds a resource in blueprint by its URI template */
-        static ResourceIteratorPair findResource(const Blueprint& blueprint,
-                                                 const Resource& resource) {
-
-            for (ResourceGroupIterator it = blueprint.resourceGroups.begin();
-                  it != blueprint.resourceGroups.end();
+        /**
+         * \brief Given a blueprint, Check if a resource already exists with the given uri template
+         *
+         * \param blueprint The blueprint which is formed until now
+         * \param uri The resource uri template to be checked
+         */
+        static bool isResourceDuplicate(const Blueprint& blueprint,
+                                        const URITemplate& uri) {
+
+            for (Elements::const_iterator it = blueprint.content.elements().begin();
+                  it != blueprint.content.elements().end();
                   ++it) {
 
-                ResourceIterator match = SectionProcessor<Resource>::findResource(it->resources, resource);
+                if (it->element == Element::CategoryElement &&
+                    SectionProcessor<Resource>::isResourceDuplicate(it->content.elements(), uri)) {
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * \brief Given list of elements, return true if none of them is a resource element
+         *
+         * \param elements Collection of elements
+         */
+        static bool isResourcePresent(const Elements& elements) {
+
+            for (Elements::const_iterator it = elements.begin(); it != elements.end(); ++it) {
+
+                if (it->element == Element::ResourceElement) {
+                    return false;
+                }
+            }
 
-                if (match != it->resources.end()) {
-                    return std::make_pair(it, match);
+            return true;
+        }
+
+        /**
+         * \brief Given list of elements, get the last resource
+         *
+         * \param elements Collection fo elements
+         */
+        static Resource lastResource(const Elements& elements) {
```

Comment:
this is not the most elegant solution, you probably want to learn about [find_if](http://www.cplusplus.com/reference/algorithm/find_if/) & predicates